exp/query/api.go

```diff
@@ -88,44 +107,182 @@ func (a *API) Search(ctx context.Context, req *SearchRequest) (*SearchResponse,
+	vectors, err := a.vectorizer.Texts(ctx, nearText, &modules.ClassBasedModuleConfig{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to vectorize the nearText query: %w", err)
+	}
+
+	// TODO(kavi): Assuming BQ compression is enabled. Make it generic.
+	bq := flatent.CompressionUserConfig{
+		Enabled: true,
+	}
+
+	// NOTE(kavi): Accept distance provider as dependency?
+	dis := distancer.NewCosineDistanceProvider()
+	index, err := flat.New(flat.Config{
+		ID:               defaultLSMVectorsCompressedBucket,
+		DistanceProvider: dis,
+		RootPath:         localPath,
+	}, flatent.UserConfig{
+		BQ: bq,
+	}, store)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
+		return nil, nil, fmt.Errorf("failed to initialize index: %w", err)
 	}
+	defer index.Shutdown(ctx)
+
+	// TODO(kavi): Here `limit` is not what you expect. It's the maxLimit.
+	// Currently `SearchByVectorDistance` api takes limit via `newSearchByDistParams(maxLimit)` which caller
+	// don't have control too.
+	certainty := 1 - threshold
+	matched_ids, distance, err := index.SearchByVectorDistance(vectors, certainty, int64(limit), nil)
 
-	if err := store.CreateOrLoadBucket(ctx, defaultLSMObjectsBucket); err != nil {
-		return nil, fmt.Errorf("failed to load objects bucket in store: %w", err)
+	opts := []lsmkv.BucketOption{
+		// lsmkv.WithStrategy(lsmkv.StrategyReplace),
+		lsmkv.WithSecondaryIndices(2),
+	}
+
+	if err := store.CreateOrLoadBucket(ctx, defaultLSMObjectsBucket, opts...); err != nil {
+		return nil, nil, fmt.Errorf("failed to vectorize query text: %w", err)
 	}
 
 	bkt := store.Bucket(defaultLSMObjectsBucket)
 
-	var resp SearchResponse
+	objs := make([]*storobj.Object, 0)
+
+	for _, id := range matched_ids {
+		key, err := indexDocIDToLSMKey(id)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to serialize ids returned from flat index: %w", err)
+		}
+		objB, err := bkt.GetBySecondary(0, key)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get object from store: %w", err)
+		}
 
-	if err := bkt.IterateObjects(ctx, func(object *storobj.Object) error {
-		resp.objects = append(resp.objects, object)
-		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("failed to iterate objects in store: %w", err)
+		obj, err := storobj.FromBinary(objB)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to decode object from store: %w", err)
+		}
+
+		objs = append(objs, obj)
 	}
 
-	return &resp, nil
+	return objs, distance, nil
+}
+
+func (a *API) loadOrDownloadLSM(ctx context.Context, collection, tenant string) (*lsmkv.Store, string, error) {
+	localPath := path.Join(a.offload.DataPath, strings.ToLower(collection), strings.ToLower(tenant), defaultLSMRoot)
+
+	// NOTE: Download only if path doesn't exist.
+	// Assumes, whatever in the path is latest.
+	// We will add a another way to keep this path upto date via some data versioning.
+	_, err := os.Stat(localPath)
```

Comment:
Do we want to lock/sync in that function to avoid 2 search in parallel triggering twice the download operation ?

Reply:
Good point. Done.